File: swift-bridge/Sources/OSAKitBridge/ScriptView.swift
import Cocoa
import OSAKit

@_cdecl("osa_script_view_new")
public func osa_script_view_new(
    _ outView: UnsafeMutablePointer<UnsafeMutableRawPointer?>,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    outView.pointee = nil
    guard osaRequireMainThread(errorOut) else {
        return OSA_MAIN_THREAD_REQUIRED
    }
    outView.pointee = osaRetain(OSAScriptView(frame: .zero))
    return OSA_OK
}

private func osaWithScriptView(
    _ viewPtr: UnsafeMutableRawPointer?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?,
    _ body: (OSAScriptView) -> Void
) -> Int32 {
    guard let viewPtr else {
        osaWriteError(errorOut, "missing OSA script view handle")
        return OSA_INVALID_ARGUMENT
    }
    guard osaRequireMainThread(errorOut) else {
        return OSA_MAIN_THREAD_REQUIRED
    }
    body(osaBorrow(viewPtr))
    return OSA_OK
}

private func osaReadScriptView<T>(
    _ viewPtr: UnsafeMutableRawPointer?,
    _ fallback: T,
    _ body: (OSAScriptView) -> T
) -> T {
    guard let viewPtr, Thread.isMainThread else { return fallback }
    return body(osaBorrow(viewPtr))
}

@_cdecl("osa_script_view_source")
public func osa_script_view_source(_ viewPtr: UnsafeMutableRawPointer?) -> UnsafeMutablePointer<CChar>? {
    osaReadScriptView(viewPtr, nil) { view in
        view.source.map(osaCString) ?? nil
    }
}

@_cdecl("osa_script_view_set_source")
public func osa_script_view_set_source(
    _ viewPtr: UnsafeMutableRawPointer?,
    _ source: UnsafePointer<CChar>?,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    osaWithScriptView(viewPtr, errorOut) { view in
        view.source = source.map { String(cString: $0) }
    }
}

@_cdecl("osa_script_view_uses_script_assistant")
public func osa_script_view_uses_script_assistant(_ viewPtr: UnsafeMutableRawPointer?) -> Bool {
    osaReadScriptView(viewPtr, false) { $0.usesScriptAssistant }
}

@_cdecl("osa_script_view_set_uses_script_assistant")
public func osa_script_view_set_uses_script_assistant(
    _ viewPtr: UnsafeMutableRawPointer?,
    _ value: Bool,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    osaWithScriptView(viewPtr, errorOut) { $0.usesScriptAssistant = value }
}

@_cdecl("osa_script_view_uses_tabs")
public func osa_script_view_uses_tabs(_ viewPtr: UnsafeMutableRawPointer?) -> Bool {
    osaReadScriptView(viewPtr, false) { $0.usesTabs }
}

@_cdecl("osa_script_view_set_uses_tabs")
public func osa_script_view_set_uses_tabs(
    _ viewPtr: UnsafeMutableRawPointer?,
    _ value: Bool,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    osaWithScriptView(viewPtr, errorOut) { $0.usesTabs = value }
}

@_cdecl("osa_script_view_tab_width")
public func osa_script_view_tab_width(_ viewPtr: UnsafeMutableRawPointer?) -> UInt64 {
    osaReadScriptView(viewPtr, 0) { UInt64(UInt(bitPattern: $0.tabWidth)) }
}

@_cdecl("osa_script_view_set_tab_width")
public func osa_script_view_set_tab_width(
    _ viewPtr: UnsafeMutableRawPointer?,
    _ width: UInt64,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    osaWithScriptView(viewPtr, errorOut) { $0.tabWidth = Int(bitPattern: UInt(width)) }
}

@_cdecl("osa_script_view_wraps_lines")
public func osa_script_view_wraps_lines(_ viewPtr: UnsafeMutableRawPointer?) -> Bool {
    osaReadScriptView(viewPtr, false) { $0.wrapsLines }
}

@_cdecl("osa_script_view_set_wraps_lines")
public func osa_script_view_set_wraps_lines(
    _ viewPtr: UnsafeMutableRawPointer?,
    _ value: Bool,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    osaWithScriptView(viewPtr, errorOut) { $0.wrapsLines = value }
}

@_cdecl("osa_script_view_indents_wrapped_lines")
public func osa_script_view_indents_wrapped_lines(_ viewPtr: UnsafeMutableRawPointer?) -> Bool {
    osaReadScriptView(viewPtr, false) { $0.indentsWrappedLines }
}

@_cdecl("osa_script_view_set_indents_wrapped_lines")
public func osa_script_view_set_indents_wrapped_lines(
    _ viewPtr: UnsafeMutableRawPointer?,
    _ value: Bool,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    osaWithScriptView(viewPtr, errorOut) { $0.indentsWrappedLines = value }
}

@_cdecl("osa_script_view_indent_width")
public func osa_script_view_indent_width(_ viewPtr: UnsafeMutableRawPointer?) -> UInt64 {
    osaReadScriptView(viewPtr, 0) { UInt64(UInt(bitPattern: $0.indentWidth)) }
}

@_cdecl("osa_script_view_set_indent_width")
public func osa_script_view_set_indent_width(
    _ viewPtr: UnsafeMutableRawPointer?,
    _ width: UInt64,
    _ errorOut: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> Int32 {
    osaWithScriptView(viewPtr, errorOut) { $0.indentWidth = Int(bitPattern: UInt(width)) }
}

File: src/script_view.rs
use core::ffi::c_void;
use std::ptr;

use crate::ffi;
use crate::private::to_cstring;
use crate::script_error::{from_swift, OsaKitError};

#[derive(Debug)]
/// Wraps the `OSAScriptView` editing view exposed by `OSAKit`.
pub struct ScriptView {
    pub(crate) raw: *mut c_void,
}

impl ScriptView {
    /// Creates an `OSAScriptView`.
    pub fn new() -> Result<Self, OsaKitError> {
        let mut raw = ptr::null_mut();
        let mut error_ptr = ptr::null_mut();
        let status = unsafe { ffi::osa_script_view_new(&mut raw, &mut error_ptr) };
        if status != ffi::status::OK {
            return Err(from_swift(status, error_ptr));
        }
        Ok(Self { raw })
    }

    /// Returns the current source text shown by this `OSAScriptView`.
    pub fn source(&self) -> Result<Option<String>, OsaKitError> {
        let ptr = unsafe { ffi::osa_script_view_source(self.raw) };
        Ok((!ptr.is_null()).then(|| crate::script_error::take_owned_c_string(ptr)))
    }

    /// Sets the source text shown by this `OSAScriptView`.
    pub fn set_source(&self, source: Option<&str>) -> Result<(), OsaKitError> {
        let source = match source {
            Some(source) => Some(to_cstring(source)?),
            None => None,
        };
        let mut error_ptr = ptr::null_mut();
        let status = unsafe {
            ffi::osa_script_view_set_source(
                self.raw,
                source.as_ref().map_or(ptr::null(), |value| value.as_ptr()),
                &mut error_ptr,
            )
        };
        if status != ffi::status::OK {
            return Err(from_swift(status, error_ptr));
        }
        Ok(())
    }

    #[must_use]
    /// Reports whether `OSAScriptView` shows the script assistant UI.
    pub fn uses_script_assistant(&self) -> bool {
        unsafe { ffi::osa_script_view_uses_script_assistant(self.raw) }
    }

    /// Enables or disables the `OSAScriptView` script assistant UI.
    pub fn set_uses_script_assistant(&self, value: bool) -> Result<(), OsaKitError> {
        let mut error_ptr = ptr::null_mut();
        let status = unsafe {
            ffi::osa_script_view_set_uses_script_assistant(self.raw, value, &mut error_ptr)
        };
        if status != ffi::status::OK {
            return Err(from_swift(status, error_ptr));
        }
        Ok(())
    }

    #[must_use]
    /// Reports whether `OSAScriptView` inserts tab characters.
    pub fn uses_tabs(&self) -> bool {
        unsafe { ffi::osa_script_view_uses_tabs(self.raw) }
    }

    /// Enables or disables tab insertion in `OSAScriptView`.
    pub fn set_uses_tabs(&self, value: bool) -> Result<(), OsaKitError> {
        let mut error_ptr = ptr::null_mut();
        let status = unsafe { ffi::osa_script_view_set_uses_tabs(self.raw, value, &mut error_ptr) };
        if status != ffi::status::OK {
            return Err(from_swift(status, error_ptr));
        }
        Ok(())
    }

    #[must_use]
    /// Returns the tab width configured on `OSAScriptView`.
    pub fn tab_width(&self) -> u64 {
        unsafe { ffi::osa_script_view_tab_width(self.raw) }
    }

    /// Sets the tab width used by `OSAScriptView`.
    pub fn set_tab_width(&self, width: u64) -> Result<(), OsaKitError> {
        let mut error_ptr = ptr::null_mut();
        let status = unsafe { ffi::osa_script_view_set_tab_width(self.raw, width, &mut error_ptr) };
        if status != ffi::status::OK {
            return Err(from_swift(status, error_ptr));
        }
        Ok(())
    }

    #[must_use]
    /// Reports whether `OSAScriptView` wraps long lines.
    pub fn wraps_lines(&self) -> bool {
        unsafe { ffi::osa_script_view_wraps_lines(self.raw) }
    }

    /// Enables or disables line wrapping in `OSAScriptView`.
    pub fn set_wraps_lines(&self, value: bool) -> Result<(), OsaKitError> {
        let mut error_ptr = ptr::null_mut();
        let status =
            unsafe { ffi::osa_script_view_set_wraps_lines(self.raw, value, &mut error_ptr) };
        if status != ffi::status::OK {
            return Err(from_swift(status, error_ptr));
        }
        Ok(())
    }

    #[must_use]
    /// Reports whether wrapped lines are indented in `OSAScriptView`.
    pub fn indents_wrapped_lines(&self) -> bool {
        unsafe { ffi::osa_script_view_indents_wrapped_lines(self.raw) }
    }

    /// Enables or disables wrapped-line indentation in `OSAScriptView`.
    pub fn set_indents_wrapped_lines(&self, value: bool) -> Result<(), OsaKitError> {
        let mut error_ptr = ptr::null_mut();
        let status = unsafe {
            ffi::osa_script_view_set_indents_wrapped_lines(self.raw, value, &mut error_ptr)
        };
        if status != ffi::status::OK {
            return Err(from_swift(status, error_ptr));
        }
        Ok(())
    }

    #[must_use]
    /// Returns the indentation width configured on `OSAScriptView`.
    pub fn indent_width(&self) -> u64 {
        unsafe { ffi::osa_script_view_indent_width(self.raw) }
    }

    /// Sets the indentation width used by `OSAScriptView`.
    pub fn set_indent_width(&self, width: u64) -> Result<(), OsaKitError> {
        let mut error_ptr = ptr::null_mut();
        let status =
            unsafe { ffi::osa_script_view_set_indent_width(self.raw, width, &mut error_ptr) };
        if status != ffi::status::OK {
            return Err(from_swift(status, error_ptr));
        }
        Ok(())
    }
}

impl Drop for ScriptView {
    fn drop(&mut self) {
        if !self.raw.is_null() {
            unsafe { ffi::osa_object_release(self.raw) };
            self.raw = ptr::null_mut();
        }
    }
}
